### PR TITLE
Feat (0x): Meta txs and Multichain

### PR DESCRIPTION
**Pull request by OjusWiZard**
Closes: #5978 

## Checklist

- [x] Add support for meta transactions in 0x decoder
- [x] Add ~~Polygon~~ Multichain support
- [x] ~~See if other chains have any chance~~

### CODE REVIEW

File: rotkehlchen/chain/evm/decoding/zerox/decoder.py [hunk 1]
```diff
@@ -146,43 +185,57 @@ def _decode_swap(
...
+        for event in context.decoded_events:
+            if ((
+                event.event_type == HistoryEventType.TRADE and
+                event.event_subtype == HistoryEventSubType.SPEND
+            ) or (
+                event.event_type == HistoryEventType.SPEND and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.address == self.flash_wallet_address
+            )) and event.location_label is not None and self.base.is_tracked(event.location_label):  # type: ignore  # it's a checksum address here
+                send_events.append(event)
```
Comment (yabirgb) [inline rotkehlchen/chain/evm/decoding/zerox/decoder.py:213]: for this last case can't you move it to the fee event and check if ` event.address == self.flash_wallet_address`. If so it is a send_event otherwise it is a fee event? This way you avoid extra checks

File: rotkehlchen/chain/evm/decoding/zerox/decoder.py [hunk 2]
```diff
@@ -146,43 +185,57 @@ def _decode_swap(
...
+
+            elif ((
+                event.event_type == HistoryEventType.TRADE and
+                event.event_subtype == HistoryEventSubType.RECEIVE
+            ) or (
+                event.event_type == HistoryEventType.RECEIVE and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.address == self.flash_wallet_address
+            )) and event.location_label is not None and self.base.is_tracked(event.location_label):  # type: ignore  # it's a checksum address here
```
Comment (yabirgb) [inline rotkehlchen/chain/evm/decoding/zerox/decoder.py:222]: Do the check for `self.base.is_tracked(event.location_label)` at the start of the loop and avoid repeating it in each case

File: rotkehlchen/chain/evm/decoding/zerox/decoder.py [hunk 3]
```diff
@@ -146,43 +185,57 @@ def _decode_swap(
...
+        """Decodes the swap event from the 0x router contract via executeMetaTransactionV2."""
+        if context.tx_log.topics[0] != META_TX_ZEROX or context.tx_log.address != self.router_address:  # noqa: E501
+            return DEFAULT_DECODING_OUTPUT
 
-        decoded_events.append(summed_send_event)
-        decoded_events.append(summed_receive_event)
-        maybe_reshuffle_events(
-            ordered_events=[summed_send_event, summed_receive_event],
-            events_list=decoded_events,
+        send_events, receive_events, fee_event = [], [], None
```
Comment (yabirgb) [inline rotkehlchen/chain/evm/decoding/zerox/decoder.py:203]: is this because you can split a trade in multiple dexes? Can you add a comment to point out this?

File: rotkehlchen/chain/evm/decoding/zerox/decoder.py [hunk 4]
```diff
@@ -65,6 +67,43 @@ def _update_send_and_receive_events(
...
+
+    def _merge_similar_events(
+            self,
+            decoded_events: 'list[EvmEvent]',
+            send_events: 'list[EvmEvent]',
+            receive_events: 'list[EvmEvent]',
+            fee_event: 'EvmEvent | None' = None,
+    ) -> None:
+        """Sum the balances of the events, updates them with the 0x values, replace them with the
+        Merged Events in decoded_events list, and maybe shuffle them with proper order."""
```
Comment (yabirgb) [inline rotkehlchen/chain/evm/decoding/zerox/decoder.py:85]: Can you in the docstring point that this is done because 0x can split the trades in multiple routes?

File: rotkehlchen/chain/evm/decoding/zerox/decoder.py [hunk 5]
```diff
@@ -65,6 +67,43 @@ def _update_send_and_receive_events(
...
+        summed_receive_event = receive_events[0] if len(receive_events) > 0 else None
+
+        if summed_send_event is not None:
+            for event in send_events[1:]:
+                summed_send_event.balance += event.balance
+                decoded_events.remove(event)
+        if summed_receive_event is not None:
+            for event in receive_events[1:]:
+                summed_receive_event.balance += event.balance
+                decoded_events.remove(event)
```
Comment (yabirgb) [inline rotkehlchen/chain/evm/decoding/zerox/decoder.py:96]: removing from a list in python irc has `O(n)` time complexity. Instead of this do a set 

```python
events_to_remove = {}
```

and add to that set the identifier of the events. Then you can do

```python
decoded_events = [event for event in decoded_events if event.identifier not in events_to_remove]
```

File: rotkehlchen/chain/evm/decoding/zerox/decoder.py [hunk 6]
```diff
@@ -65,6 +67,43 @@ def _update_send_and_receive_events(
...
             receive_event.event_subtype = HistoryEventSubType.RECEIVE
             receive_event.notes = f'Receive {receive_event.balance.amount} {receive_event.asset.symbol_or_name()} as the result of a swap via the 0x protocol'  # noqa: E501
+        if fee_event is not None:
+            fee_event.counterparty = CPT_ZEROX
+            fee_event.address = self.router_address
+            fee_event.event_type = HistoryEventType.SPEND
+            fee_event.event_subtype = HistoryEventSubType.FEE
+            fee_event.notes = f'Spend {fee_event.balance.amount} {fee_event.asset.symbol_or_name()} as a 0x protocol fee'  # noqa: E501
+
+    def _merge_similar_events(
```
Comment (yabirgb) [inline rotkehlchen/chain/evm/decoding/zerox/decoder.py:77]: I believe that the logic was hard to follow due to the name. Do you feel that is better `_merge_splitted_swap_events`

**Review by yabirgb — COMMENTED** [review]
SOme comments

File: rotkehlchen/chain/evm/node_inquirer.py [hunk 7]
```diff
@@ -1283,6 +1283,7 @@ def _process_contract_info(
         decoded_contract_info = []
         for method_name, method_value in zip(properties, output, strict=True):
             if method_value[0] is True and len(method_value[1]) != 0:
+                log.debug(f'Decoding {contract.address} with method {method_name} and value {method_value[1]!r}')  # noqa: E501
```
Comment (yabirgb) [inline rotkehlchen/chain/evm/node_inquirer.py:1286]: This needs to be removed now since it is handled in #7586

File: docs/changelog.rst [hunk 8]
```diff
@@ -2,6 +2,7 @@
 Changelog
 =========
 
+* :feature:`5978` rotki will now properly decode the swaps done via 0x protocol.
```
Comment (yabirgb) [inline docs/changelog.rst:5]: We discussed in discord that the only thing remaining is the uniswap trades in other cahins. Since uniswap is not done yet and the 0x feature is not fully supported we will add it in other PR the changelog entry once fully integrated. But leaving it here in case there are more comments in the PR

**Review by yabirgb — COMMENTED** [review]
The two failing tests are:

- remote error where we failed to query etherscan (the only node)
- known issue with eth-utils dep

File: rotkehlchen/chain/evm/decoding/zerox/decoder.py [hunk 9]
```diff
@@ -65,13 +67,54 @@ def _update_send_and_receive_events(
...
+            for event in receive_events[1:]:
+                summed_receive_event.balance += event.balance
+                events_to_remove.add(event)
+
+        self._update_send_receive_fee_events(
+            send_event=summed_send_event,
+            receive_event=summed_receive_event,
+            fee_event=fee_event,
+        )
+        # in-place filter out the events_to_remove from the decoded_events
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/zerox/decoder.py:105]:
```suggestion
        # in-place edit the decoded_events reference and filter out events_to_remove
```

File: rotkehlchen/tests/unit/decoders/test_zerox.py [hunk 10]
```diff
@@ -558,3 +626,595 @@ def test_0x415565b0_token_to_token(database, ethereum_inquirer, ethereum_account
         address=ZEROX_ROUTER,
     )]
     assert expected_events == events
+
+
+@pytest.mark.vcr()
+@pytest.mark.parametrize('ethereum_accounts', [['0xE4BeF064c912BC95C404850019909efe8D357716']])
+def test_execute_meta_transaction_v2(database, ethereum_inquirer, ethereum_accounts):
+    """Test meta transaction swaps done through via the 0x protocol router contract."""
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/decoders/test_zerox.py:634]:
```suggestion
    """Test meta transaction swaps done via the 0x protocol router contract."""
```

File: rotkehlchen/tests/unit/decoders/test_zerox.py [hunk 11]
```diff
@@ -558,3 +626,595 @@ def test_0x415565b0_token_to_token(database, ethereum_inquirer, ethereum_account
...
+        counterparty=CPT_ZEROX,
+        address=ZEROX_ROUTER,
+    )]
+    assert expected_events == events
+
+
+@pytest.mark.vcr()
+@pytest.mark.parametrize('ethereum_accounts', [['0x3951970BA92CBFff00496e8C5Ebd675cEB614773']])
+def test_execute_meta_transaction_v2_multiplex(database, ethereum_inquirer, ethereum_accounts):
+    """Test meta transaction multiplex swaps done through via the 0x protocol router contract."""
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/decoders/test_zerox.py:689]:
```suggestion
    """Test meta transaction multiplex swaps done via the 0x protocol router contract."""
```

File: rotkehlchen/tests/unit/decoders/test_zerox.py [hunk 12]
```diff
@@ -558,3 +626,595 @@ def test_0x415565b0_token_to_token(database, ethereum_inquirer, ethereum_account
...
+        counterparty=CPT_ZEROX,
+        address=ZEROX_ROUTER,
+    )]
+    assert expected_events == events
+
+
+@pytest.mark.vcr()
+@pytest.mark.parametrize('ethereum_accounts', [['0x2dcd7947973cb5CBf20e3dBF0663F566D1De9CdA']])
+def test_execute_meta_transaction_v2_flash(database, ethereum_inquirer, ethereum_accounts):
+    """Test meta transaction swaps done through via the 0x protocol using its flash contract."""
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/decoders/test_zerox.py:744]:
```suggestion
    """Test meta transaction swaps done via the 0x protocol using its flash contract."""
```

**Review by LefterisJP — APPROVED** [review]
LGTM
